SVPAlib/src/automata/sra/SRA.java

```diff
@@ -0,0 +1,1828 @@
+	}
+
+	/*
+	 * Create an automaton and removes unreachable states and only removes
+	 * unreachable states if remUnreachableStates is true and normalizes the
+	 * automaton if normalize is true
+	 */
+	public static <A, B> SRA<A, B> MkSRA(Collection<SRAMove<A, B>> transitions, Integer initialState,
+			Collection<Integer> finalStates, Collection<A> registers, BooleanAlgebra<A, B> ba, boolean remUnreachableStates, boolean normalize)
+					throws TimeoutException {
+
+		SRA<A, B> aut = new SRA<A, B>();
+
+		aut.states = new HashSet<Integer>();
+		aut.states.add(initialState);
+		aut.states.addAll(finalStates);
+
+		aut.initialState = initialState;
+		aut.finalStates = finalStates;
+		if (finalStates.isEmpty())
+			return getEmptySRA(ba);
+
+        aut.registers = registers;
+		
+        for (SRAMove<A, B> t : transitions)
+			aut.addTransition(t, ba, false);
+
+		if (normalize)
+			aut = aut.normalize(ba);
+
+		if (remUnreachableStates)
+			aut = removeDeadOrUnreachableStates(aut, ba);
+
+		if (aut.finalStates.isEmpty())
+			return getEmptySRA(ba);
+
+		return aut;
+	}
+
+	/**
+	 * Gives the option to create an automaton exactly as given by the parameters, avoiding all normalizations.
+	 * 
+	 * @throws TimeoutException
+	 */
+	public static <A, B> SRA<A, B> MkSRA(Collection<SRAMove<A, B>> transitions, Integer initialState,
+			Collection<Integer> finalStates, Collection<A> registers, BooleanAlgebra<A, B> ba, boolean remUnreachableStates, boolean normalize, boolean keepEmpty)  
+					throws TimeoutException{
+		SRA<A, B> aut = new SRA<A, B>();
+
+		aut.states = new HashSet<Integer>();
+		aut.states.add(initialState);
+		aut.states.addAll(finalStates);
+
+		aut.initialState = initialState;
+		aut.finalStates = finalStates;
+
+        aut.registers = registers;
+
+		for (SRAMove<A, B> t : transitions)
+			aut.addTransition(t, ba, true);
+
+		if (normalize)
+			aut = aut.normalize(ba);
+
+		if (remUnreachableStates)
+			aut = removeDeadOrUnreachableStates(aut, ba);
+
+		if (aut.finalStates.isEmpty() && !keepEmpty)
+			return getEmptySRA(ba);
+
+		return aut;
+	}
+	
+	// Adds a transition to the SRA
+	private void addTransition(SRAMove<P, S> transition, BooleanAlgebra<P, S> ba, boolean skipSatCheck) throws TimeoutException {
+		if (skipSatCheck || transition.isSatisfiable(ba)) {
+
+			transitionCount++;
+
+			if (transition.from > maxStateId)
+				maxStateId = transition.from;
+			if (transition.to > maxStateId)
+				maxStateId = transition.to;
+
+			states.add(transition.from);
+			states.add(transition.to);
+
+			getMovesFrom(transition.from).add((SRAMove<P, S>) transition);
+			getMovesTo(transition.to).add((SRAMove<P, S>) transition);
+		}
+	}
+
+	// ------------------------------------------------------
+	// Boolean automata operations
+	// ------------------------------------------------------
+    // FIXME: At the moment I doubt most if these work.
```

Comment:
Boolean automata operations are something we'll have to look at more down the line, as we first need to make sure the transitions and `Move` classes are robust enough.

Reply:
Indeed, the model is not closed under most Boolean operations